### PR TITLE
Adds naive parallel formatting implementation

### PR DESCRIPTION
**Pull request by TheAngryByrd**
Fix for #2624.  Does a rather naive "collect all the async and run at the end" but seems to give really promising numbers.

Test results on my machine

BenchmarkDotNet=v0.13.2, OS=Windows 11 (10.0.22000.1455/21H2)
12th Gen Intel Core i9-12900F, 1 CPU, 24 logical and 16 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2 DEBUG
  Job-XAWGZF : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2

Server=True  IterationCount=1  RunStrategy=ColdStart
|                  Method |             projects |       Mean |     Error |   StdDev |     Median | Ratio | RatioSD | Rank |       Gen0 |       Gen1 |      Gen2 | Allocated | Alloc Ratio |
|------------------------ |--------------------- |-----------:|----------:|---------:|-----------:|------:|--------:|-----:|-----------:|-----------:|----------:|----------:|------------:|
| **FormatCode_Sequentially** |                 **FAKE** | **3,872.3 ms** | **515.20 ms** | **593.3 ms** | **3,731.8 ms** |  **1.00** |    **0.00** |    **2** | **16000.0000** | **10000.0000** | **1000.0000** |   **4.55 GB** |        **1.00** |
|     FormatCode_Parallel |                 FAKE |   753.8 ms | 371.76 ms | 428.1 ms |   602.2 ms |  0.19 |    0.06 |    1 |  4000.0000 |  2000.0000 |         - |   4.55 GB |        1.00 |
|                         |                      |            |           |          |            |       |         |      |            |            |           |           |             |
| **FormatCode_Sequentially** | **FsToo(...)dling [23]** |   **862.6 ms** | **230.28 ms** | **265.2 ms** |   **792.4 ms** |  **1.00** |    **0.00** |    **2** |  **3000.0000** |  **2000.0000** |         **-** |   **1.15 GB** |        **1.00** |
|     FormatCode_Parallel | FsToo(...)dling [23] |   193.2 ms |  97.44 ms | 112.2 ms |   164.2 ms |  0.22 |    0.03 |    1 |  3000.0000 |  2000.0000 |         - |   1.15 GB |        1.00 |
|                         |                      |            |           |          |            |       |         |      |            |            |           |           |             |
| **FormatCode_Sequentially** |           **Plotly.NET** | **1,965.1 ms** | **300.60 ms** | **346.2 ms** | **1,884.8 ms** |  **1.00** |    **0.00** |    **2** |  **6000.0000** |  **3000.0000** |         **-** |   **2.29 GB** |        **1.00** |
|     FormatCode_Parallel |           Plotly.NET |   445.3 ms | 142.50 ms | 164.1 ms |   392.7 ms |  0.23 |    0.07 |    1 | 12000.0000 |  3000.0000 | 1000.0000 |    2.3 GB |        1.00 |
|                         |                      |            |           |          |            |       |         |      |            |            |           |           |             |
| **FormatCode_Sequentially** |               **fsharp** | **6,183.1 ms** | **634.73 ms** | **731.0 ms** | **6,004.4 ms** |  **1.00** |    **0.00** |    **2** | **20000.0000** | **10000.0000** | **2000.0000** |   **9.43 GB** |        **1.00** |
|     FormatCode_Parallel |               fsharp | 2,046.6 ms | 587.11 ms | 676.1 ms | 1,807.8 ms |  0.33 |    0.09 |    1 | 15000.0000 |  6000.0000 | 2000.0000 |   9.44 GB |        1.00 |
---

﻿Please verify your pull request is respecting our [Pull request ground rules](https://fsprojects.github.io/fantomas/docs/contributors/Pull%20request%20ground%20rules.html).
You can find more information in our [Contributors documentation section](https://fsprojects.github.io/fantomas/docs/contributors/Index.html).

### CODE REVIEW

File: src/Fantomas.CLI.Benchmarks/Runners.fs
Comment (TheAngryByrd) [inline src/Fantomas.CLI.Benchmarks/Runners.fs:1]: We could add this into the current benchmarks project, I was just going off the initial work.

File: src/Fantomas/Program.fs
Comment (TheAngryByrd) [inline src/Fantomas/Program.fs:91]: Converted many functions in this file to `async` and used `Async` variants of all stream/file methods.

File: src/Fantomas/Program.fs
Comment (TheAngryByrd) [inline src/Fantomas/Program.fs:288]: `reraise` doesn't work in computation expressions, this is the best work around.

However this is will stop all execution if one fails. I think we discussed that we should try as many files as we can, collect error results at the end. Is that right? @dawedawe @nojaf

File: src/Fantomas/Program.fs
Comment (TheAngryByrd) [inline src/Fantomas/Program.fs:293]: All of this `printfn` aren't threadsafe and thus we could be printing garbage on the screen. Possibly we should hold off until @dawedawe merges https://github.com/fsprojects/fantomas/pull/2693?

File: src/Fantomas/Program.fs
Comment (dawedawe) [reply src/Fantomas/Program.fs:288]: I would prefer to do as many files as we can, yes.

File: CHANGELOG.md
Comment (nojaf) [inline CHANGELOG.md:3]: As this PR will need to be targeted against the v6 branch, it is easier not to have a changelog entry. Please remove it.

I would add all entries in the v6 branch right before a release.
This avoids git conflicts when rebasing v6 onto main.

File: src/Fantomas.CLI.Benchmarks/Runners.fs
Comment (nojaf) [reply src/Fantomas.CLI.Benchmarks/Runners.fs:1]: Yeah re-using the current benchmark project make more sense to me.

File: src/Fantomas/Fantomas.fsproj
Comment (nojaf) [inline src/Fantomas/Fantomas.fsproj:5]: This should already be enabled by https://github.com/fsprojects/fantomas/blob/main/Directory.Build.props#L36

File: src/Fantomas/Program.fs
Comment (nojaf) [inline src/Fantomas/Program.fs:21]: I would introduce a `Sequential` flag instead and use parallel by default.

File: src/Fantomas/Program.fs
Comment (nojaf) [reply src/Fantomas/Program.fs:91]: Makes sense!

File: src/Fantomas/Program.fs
Comment (nojaf) [reply src/Fantomas/Program.fs:288]: Yup, I think all files should be formatted and at the end, we see which ones failed.

File: src/Fantomas/Program.fs
Comment (TheAngryByrd) [reply src/Fantomas/Program.fs]: ✔️

File: src/Fantomas/Fantomas.fsproj
Comment (TheAngryByrd) [reply src/Fantomas/Fantomas.fsproj]: ✔️

File: CHANGELOG.md
Comment (TheAngryByrd) [reply CHANGELOG.md]: ✔️

File: src/Fantomas/Program.fs
Comment (TheAngryByrd) [reply src/Fantomas/Program.fs]: Rebased onto the 6.0 branch

File: src/Fantomas/Program.fs
Comment (TheAngryByrd) [reply src/Fantomas/Program.fs:303]: Ok this is changed to keep going thru all files. Reporting setup currently to look like:

`[17:42:59 INF] Formatted: 45, Ignored : 0, Unchanged : 72, Errored: 0`

File: src/Fantomas/Program.fs
Comment (TheAngryByrd) [inline src/Fantomas/Program.fs:421]: Should match on the exception. If it's a fantomas exception then should the error message without a stacktracke. Otherwise if it's verbose then put the the whole exception in, even if it's not from fantomas.

File: src/Fantomas.CLI.Benchmarks/Runners.fs
Comment (TheAngryByrd) [inline src/Fantomas.CLI.Benchmarks/Runners.fs:19]: Should we download these repositories automatically as part of the suite?

Comment (nojaf) [reply src/Fantomas.CLI.Benchmarks/Runners.fs:19]: I have something else in mind for that. This shouldn't be part of this PR I think.

**Comment by nojaf** [comment]
@TheAngryByrd is the `--check` flag also running in parallel?

**Comment by TheAngryByrd** [comment]
> @TheAngryByrd is the `--check` flag also running in parallel?

I didn't touch it so doubtful.

**Comment by TheAngryByrd** [comment]
> > @TheAngryByrd is the `--check` flag also running in parallel?
> 
> I didn't touch it so doubtful.

Seems to already be running parallel: https://github.com/fsprojects/fantomas/blob/main/src/Fantomas/Format.fs#L118

File: src/Fantomas.Tests/Integration/ConfigTests.fs
Comment (TheAngryByrd) [inline src/Fantomas.Tests/Integration/ConfigTests.fs:55]: We discussed we didn't want to show all exception messages unless they came from specific fantomas exceptions. However, fantomas uses `failwith` in many places instead of it's own exceptions.  Should we keep this requirement, always show all the exceptions, or rework all the places with `failwith` to throw a known Fantomas exception?

**Review by TheAngryByrd — COMMENTED** [review]
1 requirement needed to be redefined but ready to review otherwise.

**Review by nojaf — CHANGES_REQUESTED** [review]
Great work Jimmy! This looks really good!

As always, I had some nitpicks but with some tweaks, we can merge this first version I think.
I think we will need to dogfood this a bit to ensure everything is what we want.

File: src/Fantomas/Program.fs
Comment (dawedawe) [reply src/Fantomas/Program.fs:22]: I'd prefer to go without it.

**Comment by nojaf** [comment]
Hi all, I made some liberal changes to this PR:

Normal logging:
image

Verbose logging:
image

Cliffnotes:
- Use `-v` for `--verbosity instead of `version`.
- Raise a specific `ParseException`.
- Set the Serilog level based on our internal flag.
- Removed the `Sequential` flag.
- Don't show `INF` when you are in `Normal` mode. MSBuild also doesn't show any specific level in their logs.

This by no means needs to be the final proposal for v6, but I think it is a step in the good direction and allows us to merge @TheAngryByrd's work.

What do you think @dawedawe @TheAngryByrd @josh-degraw @edgarfgp?

**Comment by dawedawe** [comment]
I like it a lot. Had something similar in mind but planned to wait for the merge of this one.

**Review by nojaf — APPROVED** [review]

**Pull request merged**
